src/BaseVault.sol

```diff
@@ -0,0 +1,662 @@
+            $.slot := 0x22cdba5640455d74cb7564fb236bbbbaf66b93a0cc1bd221f1ee2a6b2d0a2427
+        }
+    }
+
+    /**
+     * @notice Internal function to get the asset storage.
+     * @return $ The asset storage.
+     */
+    function _getAssetStorage() internal pure returns (AssetStorage storage $) {
+        assembly {
+            $.slot := 0x2dd192a2474c87efcf5ffda906a4b4f8a678b0e41f9245666251cfed8041e680
+        }
+    }
+
+    /**
+     * @notice Internal function to get the processor storage.
+     * @return $ The processor storage.
+     */
+    function _getProcessorStorage() internal pure returns (ProcessorStorage storage $) {
+        assembly {
+            $.slot := 0x52bb806a772c899365572e319d3d6f49ed2259348d19ab0da8abccd4bd46abb5
+        }
+    }
+
+    //// ADMIN ////
+
+    bytes32 public constant PROCESSOR_ROLE = keccak256("PROCESSOR_ROLE");
+    bytes32 public constant PAUSER_ROLE = keccak256("PAUSER_ROLE");
+    bytes32 public constant UNPAUSER_ROLE = keccak256("UNPAUSER_ROLE");
+    bytes32 public constant PROVIDER_MANAGER_ROLE = keccak256("PROVIDER_MANAGER_ROLE");
+    bytes32 public constant BUFFER_MANAGER_ROLE = keccak256("BUFFER_MANAGER_ROLE");
+    bytes32 public constant ASSET_MANAGER_ROLE = keccak256("ASSET_MANAGER_ROLE");
+    bytes32 public constant PROCESSOR_MANAGER_ROLE = keccak256("PROCESSOR_MANAGER_ROLE");
+
+    /**
+     * @notice Sets the provider.
+     * @param provider_ The address of the provider.
+     */
+    function setProvider(address provider_) external virtual onlyRole(PROVIDER_MANAGER_ROLE) {
+        if (provider_ == address(0)) {
+            revert ZeroAddress();
+        }
+        _getVaultStorage().provider = provider_;
+        emit SetProvider(provider_);
+    }
+
+    /**
+     * @notice Sets the buffer strategy.
+     * @param buffer_ The address of the buffer strategy.
+     */
+    function setBuffer(address buffer_) external virtual onlyRole(BUFFER_MANAGER_ROLE) {
+        if (buffer_ == address(0)) {
+            revert ZeroAddress();
+        }
+
+        _getVaultStorage().buffer = buffer_;
+        emit SetBuffer(buffer_);
+    }
+
+    /**
+     * @notice Sets the processor rule for a given contract address and function signature.
+     * @param target The address of the target contract.
+     * @param functionSig The function signature.
+     * @param rule The function rule.
+     */
+    function setProcessorRule(address target, bytes4 functionSig, FunctionRule calldata rule)
+        public
+        virtual
+        onlyRole(PROCESSOR_MANAGER_ROLE)
+    {
+        _getProcessorStorage().rules[target][functionSig] = rule;
+        emit SetProcessorRule(target, functionSig, rule);
+    }
+
+    /**
+     * @notice Adds a new asset to the vault.
+     * @param asset_ The address of the asset.
+     * @param active_ Whether the asset is active or not.
+     */
+    function addAsset(address asset_, bool active_) public virtual onlyRole(ASSET_MANAGER_ROLE) {
```

Comment:
the way of adding strategies (as assets - meaning we'll need to wrap every strategy?) seems not scaleable as we'll need to iterate over that list when updating total assets. also cant remove strategies?

Reply:
yeah removal needs to be added exactly

Right now, you can only toggle off the active boolean

Reply:
good point about the gas issue with addAsset should be acceptable for the start, as we'll have on the order of 10 strats at most